Initial support for matrices

FreneticScribbler opened the pull request:
Hey - love the work :D I'm in the process of integrating a VS-88A 8x8 audio matrix into HA. I've dropped the serial support I initially wrote as I'm now using ser2net to bridge the serial connection to the existing TCP endpoint, which seems to work quite okay - except for the speed, as per the forced slowdown I've had to stick in. This is one of the things I want to discuss with yourself and find a better solution for. 

The other big thing is currently I'm having to poll the full state of the matrix super often, but I haven't been able to find a smarter way to do so since a single Protocol2000 'REQUEST STATUS' instruction doesn't describe a full route - you need the following:
1. Send 'REQUEST STATUS' with output value equal to output number one is requesting status for
2. Unit replies with 'REQUEST STATUS' with output value equal to the input number that output is routed for 

Unfortunately, there's no easy way, as far as I can tell, to 'match' the two instructions from information contained within the instructions themselves. 

Apologies about some of the rogue whitespace edits, I accidentally ran a formatter from a different project over some of the files I edited and am having to manually revert. :/ But wanted to get this PR up as a draft to start the discussion over the above points.

FreneticScribbler committed:
Initial support for matrices

krohrbaugh commented on src/kesslerav/protocol2k/io.py:
```diff
@@ -13,10 +14,11 @@ class Command(IntEnum):
   ERROR = 16
   PANEL_LOCK = 30
-  QUERY_OUTPUT_STATUS = 5
+  QUERY_OUTPUT_STATUS = 6 # Query audio status so we can transparently cope with audio only matrices - for video matrices the audio routing should always match the video routing
```
I don't believe that this is correct.

Some video matrix switches support switching audio and video from two different inputs to one output. This likely needs to be `QUERY_AUDIO_OUTPUT_STATUS` and the above changed to `QUERY_VIDEO_OUTPUT_STATUS`.

FreneticScribbler replied:
Hmm. Damn. So much for that cunning plan. I'm not sure how best to identify audio only matrices then, as mine replies to, say, IDENTIFY MACHINE[1,0] (Request 1st four digits of video machine) with the first four digits of the audio machine. 

It may have to be a user override.

krohrbaugh commented:
Thanks for helping out with this! Sadly, I think it may take a bit of iteration, since I didn't have a matrix to start from, so it's likely that the existing abstractions won't be a completely smooth evolution. I hope we can work through the problems and come out the other side with an improved library. :heart:

> Apologies about some of the rogue whitespace edits, I accidentally ran a formatter from a different project over some of the files I edited and am having to manually revert.

Yeah, this makes reading the diff pretty challenging currently. 😢 

> which seems to work quite okay - except for the speed, as per the forced slowdown I've had to stick in

Unfortunately, I don't see what you're referring to in the diff.

> The other big thing is currently I'm having to poll the full state of the matrix super often

Thinking through how you want this to integrate within Home Assistant (HA) might make sense, given the protocol's limitations. For example, we could instead model this matrix as multiple `MediaSwitch` instances: one for each output. That then makes HA be able to list [8 `MediaPlayer` devices](https://www.home-assistant.io/integrations/media_player/), each of which can select a source independently. 

That could potentially allow for the constructor function (e.g., `get_tcp_media_switch`) to return either one (or multiple) `MediaSwitch` instances, depending on the number of outputs.

Another possibility is a new constructor function just for matrices that returns a `list[MediaSwitch]` or perhaps a `MediaMatrix` protocol that just wraps a `list[MediaSwitch]`.

Either way, I think we want to think more about how a matrix is a collection of switches in one device, so that we end up with an abstraction that can work with audio, and video matrices, while still being relatively straight-forward to integrate into HA.

What do you think about either the direct `list[MediaSwitch]` approach, or `MediaMatrix` being a wrapper backed by a `list[MediaSwitch]`?

FreneticScribbler commented on src/kesslerav/protocol2k/io.py:
```diff
@@ -251,33 +258,33 @@ def process(self, instructions: list[Instruction] | Instruction) -> list[Instruc
       while len(result) < 1:
-        data = conn.recv(TcpDevice.BUFFER_SIZE_BYTES)
-        responses = struct.iter_unpack(Instruction.FORMAT, data)
+        time.sleep(0.5) # Slow down to prevent 'port already in use' errors when interacting with serial over TCP TODO configurable, default off
```
@krohrbaugh this is that forced slowdown - without it my matrix was dropping quite a few commands.

I think I need to figure out how to get ser2net to add a buffer/slowdown on that side instead.

FreneticScribbler commented:
Oh totally. I'm a hacky 'just make it work' sort of coder, and your repo is nice and clean, I was very much expecting iteration :p 

I actually have tentatively implemented the matrix into HA as 8 media player entities, although linked to one unified device - maybe I misunderstand HA norms but I think this makes more sense than 8 separate _devices_. 

More than happy to switch to modelling a matrix as list[MediaSwitch] - that's a clever approach and definitely makes more sense with the eventual implementation into HA. I don't think I especially have a preference whether it's directly or through a wrapper - do you?

FreneticScribbler committed:
Switch to list[MediaSwitch] approach

Currently a breaking change - not sure if it's better to return just MediaSwitch without wrapping in a list for switches? Returning different types feels wrong to me but I can't exactly say why.

FreneticScribbler committed:
Manual revert of inadevertent formatting changes

krohrbaugh commented:
> I actually have tentatively implemented the matrix into HA as 8 media player entities, although linked to one unified device - maybe I misunderstand HA norms but I think this makes more sense than 8 separate devices.

Yeah, sorry, that's what I meant: one device with 8 entities. I always mix them up. 😊 

> More than happy to switch to modelling a matrix as list[MediaSwitch] - that's a clever approach and definitely makes more sense with the eventual implementation into HA. I don't think I especially have a preference whether it's directly or through a wrapper - do you?

I think the `list[MediaSwitch]` works best unless we find that we need additional metadata, aside from number of outputs (which would be the length of the list). Should we end up needing a way to differentiate between video and audio-only, and things like that, then the wrapper makes sense. Basically, if we could get to a constructor function that returns either one `MediaSwitch` or a list of them when the device is a matrix, that'd be a pretty clean API.

I do think it'd be good to sort out a way to differentiate between audio-only and video matrices and switches, but since both of us only have one type of device, we may have to work together on a set of commands that make that possible. But once that's sorted, I still think we could encapsulate that information in `MediaSwitch` itself, rather than in a wrapper.

FreneticScribbler commented:
So you'd rather return one MediaSwitch without being wrapped in a list, from the constructor? Ok :+1: 

Totally agreed that it would be good to be able to differentiate but I'm currently seeing no automagic way to do it - Kramer's implementation of audio only devices is likely not uniform given that mine (IIRC, this may be reversed in reality) ignores 'SET VIDEO' commands but also ignores 'QUERY AUDIO' commands, with the query information being returned, correctly, from 'QUERY VIDEO'. Weird! 

I also suspect, to be fair, that audio only devices are pretty rare in the wild! 

I think I do have a 4x4 XGA/Audio Matrix somewhere, probably need to dig that out for testing purposes...
